include/field3d.hxx

```diff
@@ -688,38 +688,40 @@ inline void checkData(const Field3D &UNUSED(f), REGION UNUSED(region) = RGN_NOBN
 
 /// Makes a copy of a field \p f, ensuring that the underlying data is
 /// not shared.
-const Field3D copy(const Field3D &f);
+Field3D copy(const Field3D &f);
 
 /// Apply a floor value \p f to a field \p var. Any value lower than
 /// the floor is set to the floor.
 ///
 /// @param[in] var  Variable to apply floor to
 /// @param[in] f    The floor value
 /// @param[in] rgn  The region to calculate the result over
-const Field3D floor(const Field3D &var, BoutReal f, REGION rgn = RGN_ALL);
+Field3D floor(const Field3D &var, BoutReal f, REGION rgn = RGN_ALL);
 
 /// Fourier filtering, removes all except one mode
 ///
 /// @param[in] var Variable to apply filter to
 /// @param[in] N0  The component to keep
 /// @param[in] rgn The region to calculate the result over
-const Field3D filter(const Field3D &var, int N0, REGION rgn = RGN_ALL);
+Field3D filter(const Field3D &var, int N0, REGION rgn = RGN_ALL);
 
-/// Fourier low pass filtering. Removes modes higher than \p zmax
+/// Fourier low pass filtering. Removes modes
+/// higher than \p zmax and optionally the zonal component
 ///
 /// @param[in] var   Variable to apply filter to
 /// @param[in] zmax  Maximum mode in Z
+/// @param[in] keep_zonal  Keep the zonal component if true
 /// @param[in] rgn   The region to calculate the result over
-const Field3D lowPass(const Field3D &var, int zmax, REGION rgn = RGN_ALL);
+Field3D lowPass(const Field3D &var, int zmax, bool keep_zonal, REGION rgn = RGN_ALL);
```

Comment:
Yes, I think that's correct (assuming you mean someone trying to pass what used to be `zmin` with those values).

`zmin > 0` isn't a bug as such, it's a misunderstanding of the arguments -- `zmin` doesn't exist, it's now `keep_zonal`. We could tighten the interface up further by changing `keep_zonal` from a `bool` to a strong typedef, e.g. `enum class LowPass { KeepZonal, RemoveZonal }` or a tagged-type like `True<KeepZonal>` (see ["Refactoring with bools" in this talk][1]). This would make it a compiletime error, which is preferable to a runtime error.

We could also provide a `bandPass` function that does only keep modes between `zmin` and `zmax`

[1]: https://youtu.be/BEt7cRl1awI?t=1702

Reply:
Added back the version that takes an `int keep_zonal`, marked deprecated and with an `ASSERT0` that the argument is either 0 or 1.